package singa;

message ClusterProto {
  optional int32 nworker_groups = 1;
  optional int32 nserver_groups = 2;
  optional int32 nworkers_per_group = 3 [default = 1];
  optional int32 nservers_per_group = 4 [default = 1];
  optional int32 nworkers_per_procs = 5 [default = 1];
  optional int32 nservers_per_procs = 6 [default = 1];

  // Used in standalone mode, one ip or hostname per line
  // For YARN or Mesos version, the processes are allocted dynamically,
  // hence no need to specify the hosts statically
  optional string hostfile = 10 [default=""];

  // servers and workers in different processes?
  optional bool server_worker_separate = 11 [default = false];

  // port number is used by ZeroMQ
  optional int32 start_port = 13 [default = 6723];
  // local workspace, train/val/test shards, checkpoint files
  required string workspace = 14;
  // relative path to workspace. if not set, use the default dir of glog
  optional string log_dir = 15 [default="/tmp"];
  // ip/hostname : port [, ip/hostname : port]
  optional string zookeeper_host = 16 [default = "localhost:2181"];
  // message size limit, default 1MB
  // optional int32 largest_message = 20 [default = 1048576];
  // optional float bandwidth = 21 [default = 100];  // MB/s

	//repeated ServerTopology server_group = 20;

  optional int32 stub_timeout = 30 [default = 5000];
  optional int32 worker_timeout = 31 [default = 5000];
  optional int32 server_timeout = 32 [default = 5000];

  // conduct updates at server side; otherwise do it at worker side
  optional bool server_update = 40 [default = true];
  // share memory space between worker groups in one procs
  optional bool share_memory = 41 [default = true];

  // bandwidth of ethernet, Bytes per second, default is 1 Gbps
  optional int32 bandwidth=50 [default=134217728];
  // poll time in milliseconds
  optional int32 poll_time=51 [default =100];
>>>>>>> SINGA-8 Implement distributed Hogwild
}

message ServerTopology {
  // group id
	required int32 id = 1;
	optional int32 sync_interval = 2;
  // neighbor group id
	repeated int32 neighbor = 3;
}
